### PR TITLE
feat(forms): use the rest operator in Validators.compose

### LINKED ISSUE #7432: Use the rest operator in Validators.compose
- **I'm submitting a feature request**
- **What is the current behavior?**

To compose several validators, we must pass an array of functions to `Validators.compose`. For example:

```
Validators.compose([Validators.required, Register.validYear, Register.validBirthYear])
```
- **What is the expected behavior?**

It would be more natural to not have to wrap the arguments into an array, and simply pass all the validators. For example:

```
 Validators.compose(Validators.required, Register.validYear, Register.validBirthYear)
```

The change that would allow this is to change the declaration of the compose() function from

```
static compose(validators: Function[]): Function;
```

to

```
static compose(...validators: Function[]): Function;
```
- **What is the motivation / use case for changing the behavior?**

It's not a big thing, of course, but it would make the API more pleasant to use.
- **Please tell us about your environment:**
- Angular version: 2.0.0-beta.7
- Browser: [all]
- Language: [TypeScript X.X | ES6/7]

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/7432

### CODE REVIEW

File: modules/@angular/forms/src/validators.ts [hunk 1]
```diff
@@ -8,15 +8,12 @@
...
 import {AsyncValidatorFn, ValidatorFn} from './directives/validators';
-import {StringMapWrapper} from './facade/collection';
-import {isPresent} from './facade/lang';
 import {AbstractControl} from './model';
 import {isPromise} from './private_import_core';
 
-function isEmptyInputValue(value: any) {
-  return value == null || typeof value === 'string' && value.length === 0;
+function isEmptyInputValue(value: any): boolean {
+  return value === null || typeof value === 'string' && value.length === 0;
```
Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: Should be == please revert

File: modules/@angular/forms/src/validators.ts [hunk 2]
```diff
@@ -117,23 +114,23 @@ export class Validators {
    * Compose multiple validators into a single function that returns the union
    * of the individual error maps.
    */
-  static compose(validators: ValidatorFn[]): ValidatorFn {
+  static compose(...validators: Array<ValidatorFn|ValidatorFn[]>): ValidatorFn {
     if (!validators) return null;
-    var presentValidators = validators.filter(isPresent);
-    if (presentValidators.length == 0) return null;
+    const presentValidators = flattenArray(validators).filter(validator => !!validator);
```
Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: Make filter part of flatten

**Comment by DzmitryShylovich** [comment]
```
Error: Error at /home/travis/build/angular/angular/modules/@angular/forms/src/validators.ts:154:32: Property 'assign' does not exist on type 'ObjectConstructor'.
```
Sounds like typings issue. Do I need to import something somewhere?

> flag: breaking change

I don't think it's a breaking change. It's still possible to add validators as an array.

File: modules/@angular/forms/src/validators.ts [hunk 3]
```diff
@@ -152,9 +149,23 @@ function _executeAsyncValidators(control: AbstractControl, validators: AsyncVali
...
-  var res: {[key: string]: any} =
+  const res: {[key: string]: any} =
       arrayOfErrors.reduce((res: {[key: string]: any}, errors: {[key: string]: any}) => {
-        return isPresent(errors) ? StringMapWrapper.merge(res, errors) : res;
+        return errors ? (<any>Object).assign({}, res, errors) : res;
       }, {});
   return Object.keys(res).length === 0 ? null : res;
 }
+
+function flattenArray(input: any[], out: Array<any> = []): Array<any> {
```
Comment (DzmitryShylovich) [inline modules/@angular/forms/src/validators.ts]: https://github.com/angular/angular/pull/12203#discussion_r82673961

Comment (alfaproject) [inline modules/@angular/forms/src/validators.ts]: Fair enough. I missed the fact that the input can't be properly typed. ):

**Comment by DzmitryShylovich** [comment]
@kara can you pls review this?

**Comment by splincode** [comment]
@mhevery What is the status of the task?

**Comment by kara** [comment]
Closing as stale

**Pull request closed without merging**